### DISCUSSION TITLE
Radiomaster ER3C-i outputs pulses and Arduino serial messages on PWM pins at power-up

### DISCUSSION THREAD

**Discussion by ken830** [ExpressLRS/ExpressLRS#2913] [Q&A] [answered]
Is this a bug or expected?

I'm returning to R/C (surface) after being mostly inactive for many years and just got a new MT12 along with some ER3C-i PWM receivers to upgrade all of my old cars. I generally have steering servo connected to CH1 and ESC to CH2.

I have an old AE TC4 FT that has an old Novak GTB ESC plugged into Channel 2 of the receiver and every time I power-up the vehicle, I get a short burst of full throttle. This happens whether the transmitter is on or not. Fail-safe settings don't matter. This doesn't happen when Transmitter disconnects (power off) -- the Rx goes into failsafe setting as expected.

I hooked up my oscilloscope to look at what's happening and I see this on Channel 2 PWM pin:

![RigolDS139](https://github.com/user-attachments/assets/4e5bcf18-f796-4566-a8c8-8a5423fcbbb0)

Scope-Ch1 (Yellow) is power to the Rx. Scope-Ch3 (Pink)) is Rx-Ch2-PWM

The PWM pin is held high (and toggles low) for ~172ms. This is what is causing the ESC to burst at full throttle momentarily at power-up. Zooming in, I recognize a consistent-width (13.345µs) square waves and it looks like serial data to me so I decode it at 75kHz, 8, N, 1 and it appears to be an Arduino console message. I'm not that familiar with Arduino, but this is what it decodes to:

![RigolDS144](https://github.com/user-attachments/assets/0250d99c-42da-4690-82f5-ba743a9ac58d)

I hook up a USB-to-TTL adapter and load up a terminal to easily copy-paste the output:

```
 ets Jan  8 2013,rst cause:2, boot mode:(3,6)                                   
                                                                                
load 0x4010f000, len 3424, room 16                                              
tail 0                                                                          
chksum 0x2e                                                                     
load 0x3fff20b8, len 40, room 8                                                 
tail 0                                                                          
chksum 0x2b                                                                     
csum 0x2b                                                                       
v00086240                                                                       
~ld                                                                             
rf cal sector: 252                                                              
freq trace enable 0                                                             
rf[112] : 00
```

I do see the "boot mode" change occasionally to `(3,0)`, `(3,3)`, etc...

Now my question is... is this normal or is this a crash/error output? I don't know the design of the Rx, but I imagine this is a default UART pin on the microcontroller (ESP32??) or something? Is this happening even before the ExpressLRS code is running? I see this on both ExpressLRS 3.4.3 and 3.5.0-RC1 releases.

For Channel 1, the PWM also does something funky at power up for the first 150ms. This one looks like a 25.97MHz clock.

![RigolDS146](https://github.com/user-attachments/assets/f959de58-fe1b-410f-8238-c8226fc4b87d)

Zooming in:

![RigolDS147](https://github.com/user-attachments/assets/a21eb568-43e6-4413-aaf4-03e6fae439b1)

Channel 3's PWM looks the least-weird, but it still holds the line high for 150ms (could it be the Rx line of the UART?):

![RigolDS151](https://github.com/user-attachments/assets/ea13f111-f94e-4214-acf9-674d84966778)

At least this one doesn't toggle so I just moved the ESC to the Channel 3 output and re-configured the Rx Channel 3 output to Channel 2 input. My Novak ESC seems to power-up quietly on this channel, but I still see this as a work-around.

**Comment by pkendall64** [comment, marked as answer]
These messages are output by the ROM on chip on the ESP8285 and there is no way of disabling them. The main problem is that there just isn't enough pins to be able to avoid using the default serial pins as PWM pins. The ESP8285 also does other funky things on some pins at boot time as well, like outputting a PWM signal and some other glitches.

ESP32 based receivers do not have this problem as we use different pins for the PWM pins. The newer ESP chips, like the ESPS32C3, which is starting to be used in newer PWM receivers also have the ability to disable the ROM output messages so will be even better for PWM receivers.

**Reply by ken830** [reply]
Thank you for clearing that up for me! My workaround of using CH3 isn't too bad.

I have an ER5C-i coming (along with another ER3C-i) to play and experiment with. Do you know what is the ER5C-i based on?

How can I tell which receivers are based on which ESP chips? Are there schematics for any of these? I looked in [ExpressLRS-Hardware](https://github.com/ExpressLRS/ExpressLRS-Hardware) and didn't see anything.

**Reply by pkendall64** [reply]
The ER5 is also 8285 based but the other channel pins should be ok.
You can tell which receivers use what chips by looking at the `targets.json` file in the targets repository. They all have a `platform` which states which chip is used.
In the future this problem will be resolved as the replacement chip for the 8285 is the ESP32C3 which uses a hardware module in the ESP chip to do the PWM output, so it's smoother, more accurate and much less jitter.

**Reply by ken830** [reply]
I'm a bit surprised to hear that the two additional PWM pins on the ER5C-i are okay... I looked at the **targets.json** file and saw that both the ER3C-i and ER5C-i use the same **Generic 2400 PWMP5.json** layout file that specifies `"pwm_outputs": [0,1,3,9,10]` and ER3C-i overrides that with `"pwm_outputs": [0,1,3]`. Wonder why they didn't choose, say, 9 & 10 to be CH1 and CH2 on the ER3C-i. Do all targets require UART Tx/Rx be exposed?

Do some of those numbers in the JSON represent the IO port number? The ESP8285 datasheet seems to suggest it only has 4 hardware PWMs: On IO12, IO15, IO14, & IO4. But they don't match any of the numbers in the JSON file (0,1,3,9,10). However, things like MISO (12), MOSI (13),  & SCK (14) on the HSPI interface seem to match perfectly to IO12, IO13, and IO14 -- and these use up some of the PWM pins. Does this mean that the PWM outputs are all software PWMs? And this confirms my original question: CH2 is on IO1, which is UART0's Tx and CH3 is on IO3, which is UART0's Rx. 

Maybe I'll look for more documentation to read instead of repeatedly bugging you here. Thanks for pointing me in the right direction.

Looking forward to the ESP32C3, but the current ESP8285 receivers are miles better than what they are replacing in my models: Synthesized FM receivers — That's how long I've been away from R/C!

**Reply by pkendall64** [reply]
We needed to have pins 1 & 3 exposed so users can flash the receiver as these are the UART pins use by the ROM for flashing.
Yes, the numbers in the JSON are the GPIO numbers. The datasheet only shows what it defined when the chip starts i.e. what the ROM sets as defaults. Any pin can be use for any function with some caveats i.e. ADC is only on one pin.

**Reply by ken830** [reply]
Thank you so much for your time in responding!

When I get my ER5C-i, I'll play with it a bit, and maybe move ESC and steering servo to Channels 4 & 5 if they are clean. Right now, the ESC behaves okay on CH3 with the signal held high at power up, but steering swings to one side (some servos more than others) on power up.

I will also consider designing a simple PCB to "fix" the issue with a quad bilateral switch IC and a RC delay circuit to block the PWM pins for the first ~250ms or so.

**Comment by m13862139345-bit** [comment]
hello, has the new firmware updates addressed the run away issues yet?

**Reply by m13862139345-bit** [reply]
if not, can manually setting arm on channel 5 (safety for throttle)solve this issue
